### PR TITLE
make rbac authorizer use rule comparison, not covers

### PR DESCRIPTION
**Pull request by deads2k**
Updates the rbac authorizer to use rule comparisons and not a covers check to perform authorization.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32813)

### CODE REVIEW

**Comment by deads2k** [comment]
@ericchiang ptal

I'll do more work later to have this operate against a cache, but this ought to be enough to get started.

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 1]
```diff
@@ -100,104 +92,103 @@ type ClusterRoleBindingLister interface {
...
-	}
-	return policyRules, nil
+	return policyRules, utilerrors.NewAggregate(errorlist)
 }
 
-func appliesTo(ctx api.Context, subjects []rbac.Subject) (bool, error) {
-	user, ok := api.UserFrom(ctx)
-	if !ok {
-		return false, fmt.Errorf("no user data associated with context")
+// GetRoleReferenceRules attempts resolve the RoleBinding or ClusterRoleBinding.
```
Comment (ericchiang) [inline pkg/apis/rbac/validation/rulevalidation.go]: s/attempts resolve/attempts to resolve/g

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 2]
```diff
@@ -100,104 +92,103 @@ type ClusterRoleBindingLister interface {
...
+		return subject.Name == rbac.UserAll || user.GetName() == subject.Name
+
 	case rbac.GroupKind:
-		return has(user.GetGroups(), subject.Name), nil
+		return has(user.GetGroups(), subject.Name)
+
 	case rbac.ServiceAccountKind:
-		if subject.Namespace == "" {
-			return false, fmt.Errorf("subject of kind service account without specified namespace")
+		// default the namespace to namespace we're working in if its available.  This allows rolebindings that reference
```
Comment (ericchiang) [inline pkg/apis/rbac/validation/rulevalidation.go]: :+1: to the improvement

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 3]
```diff
@@ -37,34 +37,26 @@ type AuthorizationRuleResolver interface {
 	// GetEffectivePolicyRules returns the list of rules that apply to a given user in a given namespace and error.  If an error is returned, the slice of
```
Comment (ericchiang) [inline pkg/apis/rbac/validation/rulevalidation.go]: Doc strings needs updating.

**Review by ericchiang — APPROVED** [review]
two doc updates, but otherwise lgtm

**Comment by deads2k** [comment]
@k8s-bot test this: issue #32770

**Comment by deads2k** [comment]
@k8s-bot test this: issue #32770

**Pull request merged**